(low power) UARTE implementation

timokroeger opened the pull request:
Outstanding work
* Example code
* Make it generic so it works with UART0, UART1, UARTx…
* Add and review documentation
* Is `stop()` on the `ReceiveFuture` best practice?
* Allow concurrent send and receive, `&mut self` prevents that. Possible solution: Split into RX and TX parts, each with `&mut self`
* Coexistance with current uarte code (especially the ISR). Maybe there is a better way than cargo features to select between them.

xoviat commented:
May I ask what the difference is between this and the existing uarte?

Dirbaio commented:
@xoviat the existing uarte (now renamed to `BufferedUarte`) is designed so it can implement the `AsyncBufRead` and `AsyncWrite` traits. These traits are intended to work like a "pipe" or "socket" in a traditional OS. 

The main consequence of that is that it has internal tx/rx buffers.

Writing copies the data to the tx buffer and returns immediately (or blocks until there's space in the buffer). The data in the tx buffer is sent in the background. Same for receiving: it reads data that's already in the buffer, and the uart is constantly receiving.

This new uart is the more "classic" version where you do "read" and "write" operations that go directly to the uart hardware. Operations don't return until the hardware has finished doing them. There's no internal buffer. This is the async version of the embedded_hal uart trait.

Depending on what you're doing, one is better than the other, so we decided to have both in embassy

Dirbaio commented on embassy-nrf/src/low_power_uarte.rs:
```diff
@@ -0,0 +1,302 @@
+                .instance
+                .tasks_stoprx
+                .write(|w| unsafe { w.bits(1) });
+            self.uarte.instance.events_rxstarted.reset();
```
I think it's necessary to do a blocking wait for ENDRX here. The DMA hardware may still write to RAM after triggering `tasks_stoprx`.

timokroeger replied:
That one is a good catch!
In most cases ENDRX has occurred already because the future will be dropped after the transfer has finished. But we abort the transfer we run into the exact problem you stated.

**Case 1: Drop after receive has finished**  
ENDRX has already been cleared by the ISR and will not be set again. But no DMA is ongoing -> Safe to drop

**Case 2: Drop the future during ongoing reception**  
STOPTX aborts the reception and the DMA transfer is still ongoing. We need to wait for ENDRX to be set before we safely drop the future.

The bust solution I could come up so far is to extend `Signal` with some `is_pending()` method and use that like: `while STATE.rx_done.is_pending() {}`
For Case 1 it is not pending and we drop immediately.
For Case 2 it blocks until the ISR triggers the signal

Dirbaio replied:
sounds good! Perhaps a `signal.blocking_wait()` method would be better instead of `signal.is_pending()`.

Dirbaio commented on embassy-nrf/src/low_power_uarte.rs:
```diff
@@ -0,0 +1,302 @@
+                .write(|w| unsafe { w.maxcnt().bits(len as _) });
+
+            trace!("startrx");
+            uarte.events_rxto.reset();
```
I think this is not needed?

Dirbaio commented on embassy-nrf/src/low_power_uarte.rs:
```diff
@@ -0,0 +1,302 @@
+                .instance
+                .tasks_stoptx
+                .write(|w| unsafe { w.bits(1) });
+            self.uarte.instance.events_txstarted.reset();
```
Same as the other: I think this needs a blocking wait.

Dirbaio commented on embassy-nrf/src/low_power_uarte.rs:
```diff
@@ -0,0 +1,302 @@
+        txd: &mut Pin<Output<PushPull>>,
+        parity: Parity,
+        baudrate: Baudrate,
+    ) {
```
It'd be cool to take the same `Pins` struct as `BufferedUarte`. Perhaps moving `BufferedUarte` to the `uarte` module? so that the module structure is `uarte::Uarte, uarte::BufferedUarte, uarte::Pins`

timokroeger replied:
Agreed. I changed the method to take the nrf-hal Pin structure. That one can also be used by `BufferedUarte`.

Dirbaio commented on embassy-nrf/src/low_power_uarte.rs:
```diff
@@ -0,0 +1,302 @@
+    /// Sets the baudrate, parity and assigns the pins to the UARTE peripheral.
+    // TODO: Make it take the same `Pins` structs nrf-hal (with optional RTS/CTS).
+    // TODO: #[cfg()] for smaller device variants without port register (nrf52810, ...).
+    pub fn configure(
```
if is a separate method, it's possible to create an instance and use it directly without configuring first. `BufferedUarte` takes all the config in `new` for this reason.

It can be useful to have a `configure` method to change the config, but `new` must leave the instance in a ready-to-use state

timokroeger replied:
nrf-hal and `BufferedUarte` take ownership of the pins.
I opted for this solution (pins by ref) because I need to time multiplex the UART peripheral between multiple pins.

Looking at the datasheet I see that the peripheral can indeed operate without an explicit call to `configure()`.
It then uses the default values of 250000bps and no parity.

Anyway, I can just mirror the parameters to the `new()` function and call `configure()` from there if you prefer that :)

Dirbaio replied:
It seems to me very strange to rely on the defaults. And I think for the pins it'll have them all disconnected. It doesn't make sense to allow the user to use the uart in that broken state.

Dirbaio commented:
@timokroeger Thanks for the PR! It looks very good!

> Allow concurrent send and receive, &mut self prevents that. Possible solution: Split into RX and TX parts, each with &mut self

The current implementation is already useful for many use cases. We can merge this PR without that feature and add it later, if you want.

> Coexistance with current uarte code (especially the ISR). Maybe there is a better way than cargo features to select between them.

Check out #7. It implements "owned IRQs", allowing the user to pass the IRQ to their driver of choice and not having conflicts between drivers! :)

Also, I've renamed the existing uarte to `BufferedUarte`. Could you rename this one to `Uarte`? It's the "standard" uarte, I think it deserves that name :)

timokroeger commented:
Thank you for the review. I’ve reworked most comments and left some notes where I’m not sure on how to proceed.

The PR now includes an crude example which is very limited because using UARTE without allocator is not nice.
I will improve the examples when `BitPool` from the `net` branch is available.

The last item on the list is making the interface generic for all UARTE instances.
I’ve looked at all the other modules and each one seems to use a different pattern to share state and PAC instance with the ISR.
Do you have any recommendations off the top of your head?

Dirbaio commented on embassy-nrf/src/uarte.rs:
```diff
@@ -0,0 +1,335 @@
+        uarte
+            .intenset
+            .write(|w| w.endtx().set().txstopped().set().endrx().set().rxto().set());
+        // TODO: Set interrupt priority?
```
Some other drivers used to set a hardcoded priority, I've removed that. 
User can set the priority on the irq before passing it to us if they care about it.

Dirbaio commented on examples/src/bin/uart.rs:
```diff
@@ -8,22 +8,21 @@ use example_common::*;
 #[task]
 async fn run() {
-    let p = unwrap!(embassy_nrf::pac::Peripherals::take());
+    let p = unsafe { embassy_nrf::pac::Peripherals::steal() };
```
why steal?

timokroeger replied:
Already taken in `main` to init RTC for the executor

Dirbaio replied:
Ah, right. You can pass it as a param to `run()`. It's nice to minimize unsafe usage in examples

timokroeger replied:
That’s nifty! Didn’t know it is possible.

Dirbaio reviewed (changes requested):
> I’ve looked at all the other modules and each one seems to use a different pattern to share state and PAC instance with the ISR. Do you have any recommendations off the top of your head?

Yeah, the best way to do that is still an unsolved problem for me :sweat_smile:. I believe the way existing drivers do it can be simplified.

In this case, something like this could work. State only has signals, so it's usable with a `&`.
```rust
pub trait Instance: Deref<Target = uarte0::RegisterBlock> + Sized + private::Sealed {
    type Interrupt: OwnedInterrupt;

    #[doc(hidden)]
    fn state() -> &'static State;
}

// each instance has an impl and its `static State`.
```

There are some concerns left:

- I really think `new` should take all the configuration. If you need to switch the uarte from a set of pins to another at runtime, wouldn't a `free()` method work, like the Uart in nrf-hal? You deconstruct the uart, then construct it again with the other pins.

- As discussed on matrix, let's remove embedded-dma! :) Take borrowed slices, then make `new` unsafe with a doc comment explaining that users must not forget futures.

- Blocking waits on both drops are still missing.

If you take care of that, I'm pretty sure this'll be good to go! :D

timokroeger commented:
Squashed commits for clarity.
Fixed everything but the `new()` / `configuration()` pattern.

I use a single UARTE instance for multiple buses like:
``` rust
struct Rs485<'a> {
    pins: uarte::Pins,
    uarte: &'a RefCell<Uarte<pac::UARTE0>>,
    // ...
}

struct Modem<'a> {
    pins: uarte::Pins,
    uarte: &'a RefCell<Uarte<pac::UARTE0>>,
    // ...
}

impl<'a> Modem<'a> {
    // Similar for Rs485 but with driver enable
    async fn send(&mut self, ...) {
        let mut uarte = self.uarte.borrow_mut();
        uarte.configure(&mut self.uarte_pins, Self::PARITY, Self::BAUDRATE);
        uarte.send(...).await;
    }
}

fn main() {
    // ...
    let uarte = uarte::Uarte::new(p.UARTE0, interrupt::take!(UARTE0_UART0));
    let uarte = RefCell::from(uarte);

    let mut rs485 = Rs485::new(&uarte, rs485_pins, ...);
    let mut modem = Modem::new(&uarte, modem_pins, ...);

    loop {
        rs485.send().await;
        modem.send().await;
    }
}

```
I experimented a bit with alternative ways but did not really find a neat solution.

1. Keep current API.  
  Pro: Fits my use case :P  
  Con: User may forget to call `configure()` and use `Uarte` without configuration or connected pins.
2. Add type state so that `send()` / `receive()` is only available after configuration
3. Move PAC instance, interrupt and pins are into `Uarte` with `new()`. Move the same stuff out with `free()`.

While 3. sounds like the cleanest solution it makes the user side really complicated and I have not found a way to have nice encapsulation like in the example above.

Dirbaio commented:
There's another issue with the `configure` api: It takes `&mut Pins`, so the borrow only lasts for the duration of the `configure` call. This means the user can configure the uart with some pins, then afterwards use them for something else concurrently, for example SPI. This is definitely unsound.

I believe the only safe API is `new`+`free` (number 3 from your comment).

With the new+free API, your usecase can be achieved without unsafe like this:

```rust
struct Resources {
    irq: interrupt::UARTE0_UART0Interrupt,
    uarte: UARTE0,
}

struct Rs485<'a> {
    resources: &'a Cell<Option<Resources>>,
    pins: Option<uarte::Pins>,
}

struct Modem<'a> {
    resources: &'a Cell<Option<Resources>>,
    pins: Option<uarte::Pins>,
}

impl<'a> Modem<'a> {
    async fn send(&mut self, ...) {
        // Take the resources, replacing with None
        // unwrap will panic if resources already taken
        let resources = self.resources.replace(None).unwrap(); 
        let pins = self.pins.take().unwrap();

        // Construct the uart
        let mut uarte = uarte::Uarte::new(resources.uarte, resources.irq, pins, Self::PARITY, Self::BAUDRATE);

        // Use it
        uarte.send(...).await;

        // Deconstruct it
        let (uarte, irq, pins) = uarte.free();

        // Put back the resources
        self.resources.set(Some(Resources { uarte, irq }));
        self.pins = Some(pins);
    }
}

fn main() {
    let resources = Cell::new(Resources {
        uarte: p.UARTE0,
        irq: interrupt::take!(UARTE0_UART0),
    })

    let mut rs485 = Rs485::new(&resources, rs485_pins, ...);
    let mut modem = Modem::new(&resources, modem_pins, ...);

    loop {
        rs485.send().await;
        modem.send().await;
    }
}
```
